synapse/util/caches/lrucache.py

```diff
@@ -82,19 +93,241 @@ def enumerate_leaves(node, depth):
                 yield m
 
 
+P = TypeVar("P")
+LN = TypeVar("LN", bound="_ListNode")
+
+
+class _ListNode(Generic[P]):
```

Comment:
This thing is now a completely generic DLL implementation. I would argue strongly in favour of moving it out of the `caches` package.